src/pyhf/cli/patchset.py

```diff
@@ -0,0 +1,161 @@
+"""The pyhf spec CLI subcommand."""
+import logging
+
+import click
+import json
+import sys
+import jsonpatch
+
+from ..workspace import Workspace
+from .. import utils
+
+logging.basicConfig()
+log = logging.getLogger(__name__)
+
+
+@click.group(name='patchset')
+def cli():
+    """Operations involving patchsets."""
+
+
+@cli.command()
+@click.argument('patchset', default='-')
+@click.option(
+    '--name', help='The name of the patch to extract.', default=None,
+)
+@click.option(
+    '--output-file',
+    help='The location of the output json file. If not specified, prints to screen.',
+    default=None,
+)
+@click.option(
+    '--with-metadata/--without-metadata',
+    default=False,
+    help="Include patchset metadata in output.",
+)
+def extract(patchset, name, output_file, with_metadata):
+    """
+    Extract a patch from a patchset.
+
+    If the patchset does not contain the name, will exit with code 1.
+
+    Returns:
+        jsonpatch (:obj:`list`): A list of jsonpatch operations to apply to a workspace.
+    """
+    with click.open_file(patchset, 'r') as fstream:
+        patchset_obj = json.load(fstream)
+
+    utils.validate(patchset_obj, 'patchset.json')
+
+    extracted_patch_obj = None
+    for patch in patchset_obj['patches']:
+        if patch['name'] == name:
+            extracted_patch_obj = patch
+            break
+
+    if not extracted_patch_obj:
+        click.echo(f"There is no patch by the name '{name}' in this patchset.")
+        sys.exit(1)
+
+    if with_metadata:
+        result = extracted_patch_obj
+        result['metadata'].update(patchset_obj['metadata'])
+    else:
+        result = extracted_patch_obj['patch']
+
+    if output_file:
+        with open(output_file, 'w+') as out_file:
+            json.dump(result, out_file, indent=4, sort_keys=True)
+        log.debug("Written to {0:s}".format(output_file))
+    else:
+        click.echo(json.dumps(result, indent=4, sort_keys=True))
+
+
+@cli.command()
+@click.argument('background-only', default='-')
+@click.argument('patchset', default='-')
+@click.option(
+    '--name', help='The name of the patch to extract.', default=None,
+)
+@click.option(
+    '--output-file',
+    help='The location of the output json file. If not specified, prints to screen.',
+    default=None,
+)
+def apply(background_only, patchset, name, output_file):
```

Comment:
Also getting an error message when I use this function:

```bash
pyhf patchset apply --name C1N2_Wh_hbb_700_400 BkgOnly.json patchset.json
```

outputs:

```
Traceback (most recent call last):
  File "/anaconda3/bin/pyhf", line 8, in <module>
    sys.exit(cli())
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/anaconda3/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/anaconda3/lib/python3.6/site-packages/pyhf/cli/patchset.py", line 115, in apply
    if patch['name'] == name:
KeyError: 'name'
```

Reply:
pushed a fix for it. it's `['metadata']['name']`